File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/shapes/email_template_controller.py
import logging
import inspect
from typing import Optional
from django.http import HttpRequest, HttpResponse
from django.shortcuts import render, redirect, get_object_or_404
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from .._traits.controller import Controller
from ....Models.contact.email_template import EmailTemplate
from ....Models.contact.email_template_lang import EmailTemplateLang
from ....Models.contact.user_email_template import UserEmailTemplate
from ....Models.shapes.language import Language
from ....Models.utils.utility import Utility

logger = logging.getLogger(__name__)

class EmailTemplateController(Controller):

    @classmethod
    def index(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            usr = request.user
            if usr.type not in ['super admin', 'company']:
                return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)
            templates = EmailTemplate.objects.all()
            return render(request, 'settings/company.html', {'EmailTemplates': templates})
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def create(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            return render(request, 'email_templates/create.html')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def store(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            if request.method != 'POST':
                return redirect(get_redirect_url(request, add_data='Invalid request.'))
            name = request.POST.get('name')
            if not name:
                return redirect(get_redirect_url(request, add_data='Name field is required.'))
            tpl = EmailTemplate()
            cls._set_template(tpl, name, request.user.id)
            return redirect('email_template.index')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def show(cls, request: HttpRequest) -> HttpResponse:
        return redirect(get_redirect_url(request, add_data='Permission denied.'))

    @classmethod
    def edit(cls, request: HttpRequest, email_template_id: int) -> HttpResponse:
        return redirect(get_redirect_url(request, add_data='Permission denied.'))

    @classmethod
    def update(cls, request: HttpRequest, id: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            if request.method != 'POST':
                return redirect(get_redirect_url(request, add_data='Invalid request.'))
            from_email = request.POST.get('from')
            subject = request.POST.get('subject')
            content = request.POST.get('content')
            lang = request.POST.get('lang')
            if not from_email or not subject or not content:
                return redirect(get_redirect_url(request, add_data='All fields are required.'))
            tpl = get_object_or_404(EmailTemplate, pk=id)
            tpl.from_email = from_email
            tpl.save()
            etl, _ = EmailTemplateLang.objects.get_or_create(parent_id=id, lang=lang)
            cls._set_lang(etl, subject, content)
            return redirect('manage.email.language', id=id, lang=lang)
        except EmailTemplate.DoesNotExist:
            return redirect(get_redirect_url(request, add_data='Email Template not found.'))
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def destroy(cls, request: HttpRequest, email_template_id: int) -> HttpResponse:
        return redirect(get_redirect_url(request, add_data='Permission denied.'))

    @classmethod
    def manage_email_lang(cls, request: HttpRequest, id: int, lang: Optional[str] = 'en') -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            usr = request.user
            if usr.type != 'super admin':
                return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)
            langs = Utility.languages()
            lang_name = Language.objects.filter(code=lang).first()
            etl = EmailTemplateLang.objects.filter(parent_id=id, lang=lang).first()
            if not etl:
                return redirect(get_redirect_url(request, add_data='Data not found.'))
            tpl = EmailTemplate.objects.filter(id=id).first()
            all_tpls = EmailTemplate.objects.all()
            return render(request, 'email_templates/show.html', {
                'emailTemplate': tpl,
                'languages': langs,
                'currEmailTempLang': etl,
                'EmailTemplates': all_tpls,
                'LangName': lang_name
            })
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def store_email_lang(cls, request: HttpRequest, id: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            if request.method != 'POST':
                return redirect(get_redirect_url(request, add_data='Invalid request.'))
            subject = request.POST.get('subject')
            content = request.POST.get('content')
            lang = request.POST.get('lang')
            if not subject or not content:
                return redirect(get_redirect_url(request, add_data='Subject and Content are required.'))
            etl, _ = EmailTemplateLang.objects.get_or_create(parent_id=id, lang=lang)
            cls._set_lang(etl, subject, content)
            return redirect('manage.email.language', id=id, lang=lang)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def update_status(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            if request.method != 'POST':
                return redirect(get_redirect_url(request, add_data='Invalid request.'))
            usr = request.user
            if usr.type not in ['super admin', 'company']:
                return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)
            UserEmailTemplate.objects.filter(user_id=usr.id).update(is_active=0)
            for key, val in request.POST.items():
                if key != '_token':
                    uet = UserEmailTemplate.objects.filter(user_id=usr.id, template_id=key).first()
                    if uet:
                        uet.is_active = val
                        uet.save()
            return redirect(get_redirect_url(request, success='Status successfully updated!'))
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def _set_template(cls, obj: EmailTemplate, name: str, user_id: int):
        obj.name = name
        obj.created_by = user_id
        obj.save()

    @classmethod
    def _set_lang(cls, obj: EmailTemplateLang, subject: str, content: str):
        obj.subject = subject
        obj.content = content
        obj.save()

File: _inc/laravel/app/Http/Controllers/Shapes/EmailTemplateController.php
<?php

namespace App\Http\Controllers\Shapes;

use App\Config\Constants\{
    DatabaseConstants as DC,
    EmailsConstants as EC,
    MiddlewaresConstants as MWC,
    PermissionsConstants as PMC,
    ViewsConstants as VW
};
use App\Http\Controllers\Abstracts\Controller;
use App\Models\{
    EmailTemplate,
    EmailTemplateLang,
    Language,
    UserEmailTemplate,
    Utility
};
use Illuminate\Http\{
    JsonResponse,
    RedirectResponse,
    Request,
    Response
};
use Illuminate\Support\Facades\{
    Auth,
    DB,
    Log,
    Validator,
    View as ViewFacade
};
use Illuminate\Support\Str;

use function App\Http\Controllers\Helpers\defaultPermissionDenial;
class EmailTemplateController extends Controller
{
    private const REDIRECT_BACK = 'back';
    private const SINGULAR = 'email_template';

    public function __construct()
    {
        $this->middleware([MWC::AUTH, MWC::XSS]);
    }

    public function index(): Response|RedirectResponse
    {
        $action   = __FUNCTION__;
        $cls      = static::class;
        $sig      = "$cls::$action";
        $viewPath = VW::SET . '.company';

        return $this->measureProfile($action, function () use ($sig, $viewPath) {
            Log::info("$sig start", ['user_id' => Auth::id()]);

            $user = Auth::user();
            if (!in_array($user?->type, [PMC::SA, PMC::CPN])) {
                Log::warning("$sig denied", ['user_id' => $user?->id]);
                return redirect(self::REDIRECT_BACK)->with('error', __('Permission denied.'));
            }

            $t = microtime(true);
            $templates = EmailTemplate::all();
            $this->logExecutionTime($t, "$sig::fetchTemplates", 'completed');

            $t = microtime(true);
            if (!ViewFacade::exists($viewPath)) {
                $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');
                return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            }
            $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');

            return response()->view($viewPath, compact('templates'));
        });
    }

    public function create(): Response
    {
        $action   = __FUNCTION__;
        $cls      = static::class;
        $sig      = "$cls::$action";
        $viewPath = self::SINGULAR . '.' . $action;

        return $this->measureProfile($action, function () use ($sig, $viewPath) {
            Log::info("$sig start");

            $t = microtime(true);
            if (!ViewFacade::exists($viewPath)) {
                $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');
                return response()->noContent(404);
            }
            $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');

            return response()->view($viewPath);
        });
    }

    public function store(Request $req): RedirectResponse
    {
        $action = __FUNCTION__;
        $cls    = static::class;
        $sig    = "$cls::$action";

        return $this->measureProfile($action, function () use ($req, $sig) {
            Log::info("$sig start", ['input' => $req->all()]);

            $t = microtime(true);
            $req->validate([EC::COL_TT => 'required']);
            $this->logExecutionTime($t, "$sig::validate", 'completed');

            $user  = Auth::user();
            $title = $req->input(EC::COL_TT);

            DB::beginTransaction();
            try {
                $t = microtime(true);
                $tpl = EmailTemplate::create([
                    EC::COL_TT         => $title,
                    EC::COL_FROM       => config('app.name'),
                    EC::COL_SLG        => Str::slug($title) ?? '',
                    DC::COL_TABLE_CREATOR => $user?->id,
                ]);
                DB::commit();
                $this->logExecutionTime($t, "$sig::transaction", 'completed');

                Log::info("$sig created", ['id' => $tpl->id]);
                return redirect()->route(self::SINGULAR . '.index')
                    ->with('success', __('Email Template successfully created.'));
            } catch (\Throwable $e) {
                DB::rollBack();
                Log::error("$sig failed", ['err' => $e->getMessage()]);
                return redirect()->back()->with('error', __('Unexpected error.'));
            }
        });
    }

    public function show(int|string $id): RedirectResponse
    {
        $action = __FUNCTION__;
        $cls    = static::class;
        $sig    = "$cls::$action";

        return $this->measureProfile($action, function () use ($id, $sig) {
            Log::info("$sig start", ['id' => $id]);
            return redirect(self::REDIRECT_BACK)->with('error', __('Permission denied.'));
        });
    }

    public function edit(int|string $id): RedirectResponse
    {
        $action = __FUNCTION__;
        $cls    = static::class;
        $sig    = "$cls::$action";

        return $this->measureProfile($action, function () use ($id, $sig) {
            Log::info("$sig start", ['id' => $id]);
            return redirect(self::REDIRECT_BACK)->with('error', __('Permission denied.'));
        });
    }

    public function update(Request $req, int|string $id): RedirectResponse
    {
        $action = __FUNCTION__;
        $cls    = static::class;
        $sig    = "$cls::$action";

        return $this->measureProfile($action, function () use ($req, $id, $sig) {
            Log::info("$sig start", ['id' => $id, 'input' => $req->all()]);

            $t = microtime(true);
            $req->validate([
                'from'    => 'required',
                'subject' => 'required',
                'content' => 'required',
                'lang'    => 'required',
            ]);
            $this->logExecutionTime($t, "$sig::validate", 'completed');

            $template = EmailTemplate::findOrFail($id);
            $template->update(['from' => $req->from]);

            $langData = $req->only(['lang', 'subject', 'content']);
            $langTpl  = EmailTemplateLang::firstOrNew([
                'parent_id' => $id,
                'lang'      => $langData['lang']
            ]);
            $langTpl->fill($langData)->save();

            Log::info("$sig lang saved", ['parent_id' => $id, 'lang' => $langData['lang']]);
            return redirect()->route(VW::EMLS . '.manage.language', [$id, $langData['lang']])
                ->with('success', __('Email Template successfully updated.'));
        });
    }

    public function destroy(int|string $id): RedirectResponse
    {
        $action = __FUNCTION__;
        $cls    = static::class;
        $sig    = "$cls::$action";

        return $this->measureProfile($action, function () use ($id, $sig) {
            Log::info("$sig denied", ['id' => $id]);
            return redirect(self::REDIRECT_BACK)->with('error', __('Permission denied.'));
        });
    }

    public const MNG_EM_LNG = 'manageEmailLang';
    public function manageEmailLang(int|string $id, string $lang = DC::DEFAULT_LANG): Response|RedirectResponse
    {
        $action   = __FUNCTION__;
        $cls      = static::class;
        $sig      = "$cls::$action";
        $viewPath = VW::EML_TMP . '.show';

        return $this->measureProfile($action, function () use ($id, $lang, $sig, $viewPath) {
            $user = Auth::user();
            Log::info("$sig start", ['id' => $id, 'lang' => $lang, 'user_id' => $user?->id]);

            if ($user->type != PMC::SA) {
                return redirect(self::REDIRECT_BACK)->with('error', __('Permission denied.'));
            }

            $t = microtime(true);
            $languages     = Utility::languages();
            $langName      = Language::where('code', $lang)->first();
            $template      = EmailTemplate::findOrFail($id);
            $langTemplate  = EmailTemplateLang::firstOrNew(['parent_id' => $id, 'lang' => $lang]);
            $allTemplates  = EmailTemplate::all();
            $this->logExecutionTime($t, "$sig::fetchData", 'completed');

            $t = microtime(true);
            if (!ViewFacade::exists($viewPath)) {
                $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');
                return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            }
            $this->logExecutionTime($t, "$sig::viewExistsCheck", 'completed');

            return response()->view($viewPath, compact(
                'template',
                DC::TABLE_LANGS,
                'langTemplate',
                'allTemplates',
                'langName',
                'languages'
            ));
        });
    }

    public const STR_EM_LNG = 'storeEmailLang';
    public function storeEmailLang(Request $req, int|string $id): RedirectResponse
    {
        $action = __FUNCTION__;
        $cls    = static::class;
        $sig    = "$cls::$action";

        return $this->measureProfile($action, function () use ($req, $id, $sig) {
            Log::info("$sig start", ['id' => $id, 'input' => $req->all()]);

            $t = microtime(true);
            $req->validate(['subject' => 'required', 'content' => 'required', 'lang' => 'required']);
            $this->logExecutionTime($t, "$sig::validate", 'completed');

            $data    = $req->only(['lang', 'subject', 'content']);
            $langTpl = EmailTemplateLang::firstOrNew([
                'parent_id' => $id,
                'lang'      => $data['lang']
            ]);
            $langTpl->fill($data)->save();

            Log::info("$sig upsert", ['parent_id' => $id, 'lang' => $data['lang']]);
            return redirect()->route(VW::EMLS . '.manage.language', [$id, $data['lang']])
                ->with('success', __('Email Template Detail successfully updated.'));
        });
    }

    public const UPD_STT = 'updateStatus';
    public const IDX = 'index';
    public const CRT = 'create';
    public const STR = 'store';
    public const SHW = 'show';
    public const EDT = 'edit';
    public const UPD = 'update';
    public const DEL = 'destroy';

    public function updateStatus(Request $req): RedirectResponse
    {
        $action = __FUNCTION__;
        $cls    = static::class;
        $sig    = "$cls::$action";

        return $this->measureProfile($action, function () use ($req, $sig) {
            $user = Auth::user();
            Log::info("$sig start", ['user_id' => $user?->id, 'input' => $req->all()]);

            if (!in_array($user?->type, [PMC::SA, PMC::CPN])) {
                Log::warning("$sig denied", ['user_id' => $user?->id]);
                return redirect(self::REDIRECT_BACK)->with('error', __('Permission Denied.'));
            }

            $statuses = $req->except('_token');

            $t = microtime(true);
            DB::transaction(function () use ($user, $statuses) {
                // Reset then apply toggles
                UserEmailTemplate::where('user_id', $user->id)->update([EC::COL_IA => 0]);
                collect($statuses)->each(function ($active, $tplId) use ($user) {
                    UserEmailTemplate::updateOrCreate(
                        ['user_id' => $user?->id, 'template_id' => $tplId],
                        [EC::COL_IA => (int) (bool) $active]
                    );
                });
            });
            $this->logExecutionTime($t, "$sig::transaction", 'completed');

            Log::info("$sig complete", ['user_id' => $user?->id]);
            return redirect(self::REDIRECT_BACK)->with('success', __('Status successfully updated!'));
        });
    }

    protected function _authorize(Request $req, string $perm): ?RedirectResponse
    {
        $user = $req->user();
        if (!$user?->can($perm)) {
            Log::warning('Permission denied', [
                'method'     => __METHOD__,
                'user_id'    => $user?->id,
                'permission' => $perm,
            ]);
            return defaultPermissionDenial($req, new \Exception($perm), __CLASS__ . '::authorize');
        }
        Log::info('Permission granted', ['method' => __METHOD__, 'perm' => $perm, 'user_id' => $user?->id]);
        return null;
    }
}
